File: plugins/capacitor-native-websocket/android/src/main/java/com/capacitor/nativewebsocket/TLSCertificateStore.kt
package com.capacitor.nativewebsocket

import android.content.Context
import android.content.SharedPreferences
import java.security.MessageDigest
import java.security.cert.X509Certificate

object TLSCertificateStore {
    private const val PREFS_NAME = "ai.openclaw.tls"
    private const val KEY_PREFIX = "gateway.tls."

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun loadFingerprint(context: Context, storeKey: String): String? {
        val key = KEY_PREFIX + storeKey
        val raw = prefs(context).getString(key, null)?.trim()
        return if (!raw.isNullOrEmpty()) raw else null
    }

    fun saveFingerprint(context: Context, fingerprint: String, storeKey: String) {
        val key = KEY_PREFIX + storeKey
        prefs(context).edit().putString(key, fingerprint).apply()
    }

    fun clearFingerprint(context: Context, storeKey: String) {
        val key = KEY_PREFIX + storeKey
        prefs(context).edit().remove(key).apply()
    }

    fun certificateFingerprint(cert: X509Certificate): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val hash = digest.digest(cert.encoded)
        return hash.joinToString("") { "%02x".format(it) }
    }

    fun normalizeFingerprint(raw: String): String {
        val stripped = raw.replace(Regex("(?i)^sha-?256\\s*:?\\s*"), "")
        return stripped.lowercase().filter { it in '0'..'9' || it in 'a'..'f' }
    }
}

File: plugins/capacitor-native-websocket/ios/Sources/NativeWebSocketPlugin/TLSCertificateStore.swift
import CryptoKit
import Foundation
import Security

enum TLSCertificateStore {
    private static let suiteName = "ai.openclaw.shared"
    private static let keyPrefix = "gateway.tls."

    private static var defaults: UserDefaults {
        UserDefaults(suiteName: suiteName) ?? .standard
    }

    static func loadFingerprint(storeKey: String) -> String? {
        let key = keyPrefix + storeKey
        let raw = defaults.string(forKey: key)?.trimmingCharacters(in: .whitespacesAndNewlines)
        if let raw, !raw.isEmpty { return raw }
        return nil
    }

    static func saveFingerprint(_ value: String, storeKey: String) {
        let key = keyPrefix + storeKey
        defaults.set(value, forKey: key)
    }

    static func clearFingerprint(storeKey: String) {
        let key = keyPrefix + storeKey
        defaults.removeObject(forKey: key)
    }
}

func certificateFingerprint(_ trust: SecTrust) -> String? {
    guard let chain = SecTrustCopyCertificateChain(trust) as? [SecCertificate],
          let cert = chain.first
    else {
        return nil
    }
    return sha256Hex(SecCertificateCopyData(cert) as Data)
}

func sha256Hex(_ data: Data) -> String {
    let digest = SHA256.hash(data: data)
    return digest.map { String(format: "%02x", $0) }.joined()
}

func normalizeFingerprint(_ raw: String) -> String {
    let stripped = raw.replacingOccurrences(
        of: #"(?i)^sha-?256\s*:?\s*"#,
        with: "",
        options: .regularExpression)
    return stripped.lowercased().filter(\.isHexDigit)
}
